crates/ruff_db/src/system/memory_fs.rs

```diff
@@ -238,32 +254,33 @@ impl MemoryFileSystem {
         normalized.into_utf8_path_buf()
     }
 
-    pub fn read_directory(
-        &self,
-        path: impl AsRef<SystemPath>,
-    ) -> Result<impl Iterator<Item = Result<DirectoryEntry>> + '_> {
+    pub fn read_directory(&self, path: impl AsRef<SystemPath>) -> Result<ReadDirectory> {
         let by_path = self.inner.by_path.read().unwrap();
         let normalized = self.normalize_path(path.as_ref());
         let entry = by_path.get(&normalized).ok_or_else(not_found)?;
         if entry.is_file() {
             return Err(not_a_directory());
         };
-        Ok(by_path
-            .range(normalized.clone()..)
-            .skip(1)
-            .take_while(|(path, _)| path.starts_with(&normalized))
-            .filter_map(|(path, entry)| {
-                if path.parent()? == normalized {
-                    Some(Ok(DirectoryEntry {
-                        path: SystemPathBuf::from_utf8_path_buf(path.to_owned()),
-                        file_type: Ok(entry.file_type()),
-                    }))
-                } else {
-                    None
-                }
-            })
-            .collect::<Vec<_>>()
-            .into_iter())
+
+        // Collect the entries into a vector to avoid dead-locks when the
+        // consumer calls into other file system methods while iterating over the
+        // directory entries.
+        let mut collected = Vec::new();
+
+        for (path, entry) in by_path.range(normalized.clone()..).skip(1) {
+            if !path.starts_with(&normalized) {
+                break;
+            }
+
+            if path.parent() == Some(&normalized) {
+                collected.push(Ok(DirectoryEntry {
+                    path: SystemPathBuf::from_utf8_path_buf(path.to_owned()),
+                    file_type: entry.file_type(),
+                }));
+            }
+        }
```

Comment:
nit: I don't mind returning the named type too much, but I'm not sure I see the need to change to an imperative approach here. Does doing so fix a bug? You could do this with a much smaller diff by keeping most of the code as it was:

```suggestion
        let collected = by_path
            .range(normalized.clone()..)
            .skip(1)
            .take_while(|(path, _)| path.starts_with(&normalized))
            .filter_map(|(path, entry)| {
                if path.parent()? == normalized {
                    Some(Ok(DirectoryEntry {
                        path: SystemPathBuf::from_utf8_path_buf(path.to_owned()),
                        file_type: entry.file_type(),
                    }))
                } else {
                    None
                }
            })
            .collect()
```